CHANGELOG.md

```diff
@@ -17,6 +17,34 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
 
 ## [Unreleased](https://github.com/o1-labs/o1js/compare/ed198f305...HEAD)
 
+### Breaking changes
+
+- Fixed a vulnerability in `OffchainState` where it didn't store the `IndexedMerkleTree` length onchain and left it unconstrained https://github.com/o1-labs/o1js/pull/1676
```

Comment:
What were the consequences of not constraining the length?

Reply:
the `length` only constrains where people are allowed to insert new values. insertion prevents overwriting an existing value even if the length is unconstrained though. but, people could insert at random positions instead of linearly from the start, which could make the merkle tree harder to use and break assumptions of other users of it.

what's probable worst is that people could _hide_ where they are inserting leaves into the tree. so, they update the onchain root but then no one else can reconstruct that root because they don't know where new leafs were inserted. so it would make the contract unusable.

for example, in the offchain state implementation, everyone re-builds the tree on their own, assuming new leaves are inserted at the `length`, and it being unconstrained could have enabled an actor who calls `settle()` in a malicious way to break the offchain state root unrecoverable for anyone else.

Reply:
I thought it might be nice to hash the root and length together into a single commitment, maybe also called `root` for familiarity, so that there is only one onchain state to store and people can't forget about the length

Reply:
Oh that makes sense! I was somehow under the assumption that length meant height and the height is already embedded in the witness structure and verification, but that's obviously something different.